Rework the Bug Triage handbook

xmudrii opened the pull request:
Signed-off-by: Marko Mudrinić <mudrinic.mare@gmail.com>

This PR is a proposal to rework the Bug Triage Handbook. The current state of the handbook doesn't reflect the process we had for the past 2 release cycles (1.15 and 1.16). The purpose of this rework is to make the handbook up-to-date, easier to read and understand and to have some additional information that might help both leads and shadows.

If you think I've deleted some information that might be useful, please let me know in the review comments and I'd be happy to reconsider to add it back. I'll leave some comments on the PR for points where I wasn't sure what I should do.

I would appreciate feedback on this PR. I'm going to put in on-hold for more in-depth review.
/hold

/assign @josiahbjorgaard @justaugustus @lachie83 @guineveresaenger 
cc @kubernetes/release-team

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -1,49 +1,27 @@
+
+1. Early Release: from week 0 up to mid-release cycle (week 6). *Duration: ~6-7 weeks*
+2. Mid release: from week 6 up to the code freeze (around week 9). *Duration: ~3-4 weeks*
+3. Code Freeze: from around week 9 up to the code thaw (week 11). *Duration: ~2 weeks*
```
Do we have a definition somewhere how long each stage lasts (i.e. when the code freeze starts)? For example, is week 9 a good guess here?

tpepper replied:
9's a reasonable guess.

There is https://git.k8s.io/community/contributors/devel/sig-release/release.md for a reference.

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
-
-If you are not sure (mainly priority), you can comment and ask the owner or SIG leads about the status of the issue/PR (also see [How To Escalate](#how-to-escalate) on pinging options).
-
-Other regular patterns are:
```
I will see about keeping this part.

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
-Around this point, the release is about 1 month away, and there is less and less time for PRs to make it in time.
+  - [PRs in the v1.17 milestone with the `priority/critical-urgent` label](https://github.com/kubernetes/kubernetes/pulls?utf8=%E2%9C%93&q=is%3Aopen+is%3Apr+milestone%3Av1.17+label%3Apriority%2Fcritical-urgent): `is:open is:pr milestone:v1.17 label:priority/critical-urgent`
 
-Filtering is one of the main functions of the Bug Triage role. It involves informing owners and SIG leads about the timelines and asking  whether an issue/PR at hand is relevant to the release, whether it should be prioritized or moved to another release via the use of `milestone`. This process involves a mix of multiple parameters such as:
```
This part is now more in-depth explained for each stage, i.e. what is important for each stage and when we just start removing everything. If you find anything missing please let me know.

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
````diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
+```
+Hello! This <issue|PR> has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting <insert date> (about <number of weeks> from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
 
-### Issue Categorization
````
I don't think we had Green/Yellow/Red reports for each issue in the past two cycles, but instead, we did this for in general.

guineveresaenger replied:
yes, I think clarification is good - we don't need the individual issue colors.

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
+On the day of the Code Freeze, your responsibility is to try to help contributors to get the approval on their PRs and needed label. Check [How To Escalate](#how-to-escalate) part of the document for guide how to do this.
 
-This is enforced via test-infra automation, as a means of "new release coming in a few weeks - we're only merging relevant PRs in".
+When the code freeze starts, the highest priority has the PRs which had `approved` and `lgtm` labels before the code freeze started and are in the milestones. Depending on the merge queue length, it might be proposed to hold PRs that have `approved` and `lgtm`, but are not in the milestone.
```
Query for this: [`is:pr is:open no:milestone label:approved label:lgtm`](https://github.com/kubernetes/kubernetes/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+no%3Amilestone+label%3Aapproved+label%3Algtm).

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
+#### Once the Code Freeze starts until one week later
 
-For issues/PRs that look tricky, you can always coordinate with the Release Lead and SIG/PR owners to reach a decision.
+Usually, the removal of issues/PRs doesn't start right once the code freeze starts. Depending on the CI Signal, the merge queue length, and the release team lead's decision, it might be decided to allow some more time for PRs to get merged. At this stage, `kind/bug` PRs are prioritized, but other PRs can be allowed too if there is no risk of destabilizing the release/CI Signal (such PRs require a ...
```
Do we have documented process for PR exceptions?

tpepper replied:
Exceptions are given some treatment in the release lead's [role handbook](https://git.k8s.io/sig-release/release-team/role-handbooks/release-team-lead).  It basically amounts to SIGs' discretion.

Exceptions during freeze also start to bump into cherry-picks.  The [cherry-pick review](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md#cherry-pick-review) documentation talks about the threshold for these, but again it is largely SIGs discretion.

It would be beneficial to get more clear criteria documented for example in the [Developer Guide page on release](https://git.k8s.io/community/contributors/devel/sig-release/release.md).

josiahbjorgaard replied:
I thought that we had a pretty hard stop, at least during 1.16, such that after code freeze, everything required an exception and cherry-pick in order to be merged. This paragraph says that some PRs are allowed without exception. It reads to me that PRs which don't destablize the release are allowed without exception, but if the do destablize the release they require exception. I think it is more that any PR that is not a bug requires an exception, and those will only be granted for PRs which don't destablize the release.

xmudrii replied:
@josiahbjorgaard Yes, after the code freeze ends, everything requires a cherry-pick. As far as I know, you don't need an exception at that stage, but the cherry-pick is required. However, in the code freeze, we allowed PRs to get merged even without the formal exception. For example, we had a few PRs that were not `priority/critical-urgent` but got merged:
* https://github.com/kubernetes/kubernetes/pull/82288
* https://github.com/kubernetes/kubernetes/pull/81876
* https://github.com/kubernetes/kubernetes/pull/82248 (this one is `kubeadm` related and `kind/design`, but it still got merged without an exception, at least as far as I know)

On the other side, we had this PR https://github.com/kubernetes/kubernetes/pull/78547, which required an exception because it was not marked for 1.16 before the code freeze started.

> I think it is more that any PR that is not a bug requires an exception, and those will only be granted for PRs which don't destablize the release.

I think this is how it should be. But as @tpepper said, we should document it somewhere else, like in the developer guide, and then link it here. Also, I'd like to get some opinions from @guineveresaenger on this topic too.

guineveresaenger replied:
I would like the ideal/the goal to be reflected here and approve of the wording.

We want to have this be the standard.

I agree that we could clarify in the dev guide as well - that feel less urgent to me, because we're only talking about a tiny part of the contributor/release cycle.

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
+You can still use the queries from the previous section at this stage to monitor PRs, but you should also monitor issues using similar queries.
 
-New test failures will also show up during Code Freeze and you need to make sure that these are labeled properly, get attention of the CI Signal lead, and get attention from the appropriate SIGs.
+#### Week 1 of Code Freeze until Code Thaw starts
```
```suggestion
#### Week 1 of the Code Freeze until the Code Thaw starts
```

guineveresaenger replied:
I don't think we need the the's here.

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
 
-* Prow is an alternative tool for PR visualisation that can run full GH queries:
-    * [PRs with LGTM but not 'approved' in Prow](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20label%3Algtm%20-label%3Aapproved%20milestone%3Av1.14)
+If there is no response on the issue/PR and no recent activity, the issue/PR should be removed from the milestone to reduce the burden on the next Bug Triage team. This is done by using the `/milestone clear` command.
```
This was a little bit controversial this release cycle, but on the retro, there were no concerns, so I think we might want to keep doing this.

guineveresaenger commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
-- The time commitment becomes greater through the release cycle, peaking during Code Freeze.  In the last two weeks of the release, leading up to and during Code Freeze, Shadows should expect to spend at least 5 hours and leads at least 10 hours.
+- Availability to attend the majority of Release Team (weekly) and Burndown meetings
+- Ability to follow-up on issues and PRs since around week 6 (mid-release)
+- Since week 6 to the end of the release cycle the time commitment becomes greater. Shadows should expect to spend at least 3-5 hours and leads around 5-10 hours a week.
```
s/since/from :)

xmudrii replied:
Fixed :)

Pothulapati replied:
I spent nearly 3 hours per week watching the issues. So, 3-5 should be correct :)

josiahbjorgaard replied:
I think 3-5 is good and was what I spent, as well.

guineveresaenger commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
 
-Bug Triage relates to both the Enhancements Lead and CI Signal Lead roles.  Understanding the in-bound enhancements is important during the Early Release phase as they set the themes for incoming issues and bugs.  Having an awareness on current test status is also critical, even though there is a specific lead for that area.  The [documentation for CI Signal lead role](../ci-signal) includes a li ...
+Before starting, the Bug Triage members should refer to the following guides to get familiar with used labels:
+- [the documentation for issue `kind` labels](https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label)
```
it got moved one folder deeper and the title changed: https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label

xmudrii replied:
Thanks!

guineveresaenger commented on release-team/role-handbooks/bug-triage/README.md:
````diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
 
-1. Leave a comment on the GitHub issue or PR: "This issue hasn't been updated in 3 months.  Are you still expecting to complete it for 1.17?".  It's helpful here to @ mention individuals or SIG ```-bugs``` or ```-pr-reviews``` aliases, e.g. "@kubernetes/sig-node-bugs" or "@kubernetes/sig-network-pr-reviews".
-2. Send a message to relevant SIG slack channels or mailing list about the problem: It's helpful to directly @ mention the relevant SIG Leads / Owners, and to condense multiple issues into a list, e.g. "Hey, these three issues haven't seen any attention, are they still valid for 1.17?"
-3. Message individual owners and reviewers directly via Slack and/or email (GitHub notification emails must have been filtered or missed if you're past step #1).  Individual's email addresses may be harder to find than GitHub ID, but are usually in the git commit history.  Sometimes Slack handles are hard to find.  There is no master list mapping human names to GitHub ID, email or Slack ID.  If y ...
````
I have _never_ sent anyone an email while in this role. I'm +1 to remove the email parts.

nikopen replied:
Many sig-leads do not spend much time on upstream Slack or Github, but instead have filters on their emails to occassionally check. There's a reason why mailing lists exist.

tpepper replied:
I don't think I've ever had to resort to a direct email to an individual.  But I have used the "kubernetes-sig-${foo}-leads@googlegroups.com" mailing lists.  These lists are readily discoverable and it's a more reasonable path to go to the SIG chairs instead of trying to find the email address of an individual member of the SIG, especially when the only likely usage of these escalation paths is for time critical, exceptional situations just ahead of release.

xmudrii replied:
@tpepper Do we have some document explaining how these lists work that I can link here?

tpepper replied:
The closest I think is this:
https://github.com/kubernetes/community/blob/master/sig-wg-lifecycle.md#communicate

xmudrii replied:
@tpepper I've added a note about SIG leads mailing lists, PTAL.

guineveresaenger commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
-### Sample Searches [Early]
-* Issues which could require early intervention
-  - [Issues in the v1.17 milestone which haven't been updated in a while](https://github.com/kubernetes/kubernetes/issues?utf8=%E2%9C%93&q=is%3Aissue+milestone%3Av1.17+updated%3A%3C2019-09-01+repo%3Akubernetes%2Fkubernetes+is%3Aopen+): `is:issue milestone:v1.17 updated:<2019-09-01 repo:kubernetes/kubernetes is:open`
+The Bug Triage team is using a [tracking spreadsheet](https://docs.google.com/spreadsheets/d/14DzSChatdcwKLBxshMkE9ul4jaocBJGPdR1VnrMCqPs) to track the current status of all issues and PRs targeting the release, their priorities, and to distribute the work among the Bug Triage team. 
```
not sure we want to link this sheet, since it's 1.16 specific.

Having sample issue searches in the handbook is really helpful IMO

xmudrii replied:
Removed the link.

guineveresaenger commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
+Hello! This <issue|PR> has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting <insert date> (about <number of weeks> from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
 
-### Issue Categorization
+As the PR is tagged for 1.17, is it still planned for this release?
```
+100 love this!

guineveresaenger commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
+### Code Thaw
 
-Starting a few working days before the first Release Candidate, Code Freeze labeling restrictions are lifted.  At this point, you need to make sure that three things happen:
+Starting a week before the stable release, Code Freeze labeling restrictions are lifted. At this point, you need to make sure that three things happen:
```
maybe add the specific tag for stable release?

xmudrii replied:
Can you please give me an example for this?

guineveresaenger replied:
Apologies - I thought we were talking about a pre-release here - since "stable" can mean different things, maybe jsut drop "stable" in favor of "a week before the release target date".

xmudrii replied:
This is now fixed.

xmudrii commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -55,233 +33,208 @@ Bug Triage has a lower time requirement than most roles at the beginning of the
+#### Updating the Bug Triage Spreadsheet
 
-  - [k org-wide milestone enhancements](https://github.com/search?q=org%3Akubernetes+is%3Aissue+is%3Aopen+milestone%3Av1.17+label%3Akind%2Ffeature)`is:issue is:open milestone:v1.17 label:kind/feature org:kubernetes`
+The Bug Triage Spreadsheet is updated manually. The commands for updating the spreadsheet can be found in the `Kubernetes Github Commands` menu.
```
I added some more context, PTAL.

tpepper commented on release-team/role-handbooks/bug-triage/README.md:
```diff
@@ -1,49 +1,27 @@
-Bug Triage Handbook
-==============================
-
-- [Bug Triage Handbook](#Bug-Triage-Handbook)
```
For a large page it's useful, especially when there's different phases of work across time.

On a big page like this one I use 'pandoc' to regenerate a table of contents and copy paste it in, eg:
`pandoc -t markdown_github --toc -s README.md`

idealhack replied:
Agreed it's useful for handbooks. You can try https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one if you use VS Code.

xmudrii replied:
The table of contents is now back! :100:

josiahbjorgaard commented:
In response to @tpeppers question about removing code thaw, I don't remember thinking about the phase as code thaw during the last release cycle. I think that support period is a better reference. How common is the scenario referenced in the current state of the manual where PRs are merged without the knowledge of the release team? I don't think this happened on the last cycle.

tpepper commented:
> How common is the scenario referenced in the current state of the manual where PRs are merged without the knowledge of the release team? I don't think this happened on the last cycle.

This is an exceptional thing and low volume, but I've seen it in multiple releases in the past year and a half.

xmudrii commented:
@tpepper @josiahbjorgaard @guineveresaenger I have addressed the most of the comments, PTAL. I'd leave the code thaw changes for a follow-up PR. I can work on that one too, but I'd finish this PR and then start discussion about that in another. Although, I agree that we should drop references to it.

guineveresaenger commented:
/lgtm
/approve

Let's merge this and iterate re: code thaw (I have some changes to make in the lead handbook as well).

guineveresaenger commented:
/hold cancel
